### PR TITLE
[SVE] Add support for representing and creating buffer-level predicates

### PR DESCRIPTION
**Pull request by lhutton1**
### Representation
This commit extends `BufferLoad` and `BufferStore` to accept a predicate mask argument indicating which lanes in a vectorized buffer load/store should be read/written.

As a simple example, we can load all lanes:
```
tir.BufferLoad(buf, [tir.Ramp(0, 1, 8)], predicate=tir.Broadcast(1, 8))
```

Or disable loading all lanes:
```
tir.BufferLoad(buf, [tir.Ramp(0, 1, 8)], predicate=tir.Broadcast(0, 8))
```

In TVMScript, buffer loads and stores are currently displayed using a "short-hand" notation e.g. `A[0:4]`, but there was no clear path for extending this notation to support predicates. Therefore, the vload/vstore notation is used e.g. `A.vload([T.Ramp(0, 1, 4)], predicate=...)`. The TVMScript printer falls back to the vload/vstore notation whenever predicates are specified.

### Creation
Buffer-level predication becomes more motivating when combined with the `tir.get_active_lane_mask` intrinsic. It can be used to mask off lanes when the vectorized axis is not divisible by the vector length. A detailed example and rationale can be found in the [RFC](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0104-scalable-vectors-in-tir.md#predication).

Predicated buffer load/stores are created in the `VectorizeLoop` pass via `TryPredicateBufferAccesses`. This pass aims to convert block-level predicates e.g.
```
for i_0 in T.serial(4):
    for i_1 in T.vectorized(4):
        if i_0 * 4 + i_1 < 14:
            B[i_0 * 4 + i_1] = A[i_0 * 4 + i_1] + 1.0
```
to buffer-level predicates, e.g.
```
for i_0 in T.serial(4):
    predicate = T.get_active_lane_mask("int1x4", i_0 * 4, 14)
    A_load = T.meta_var(A.vload([T.Ramp(i_0 * 4, 1, 4)], predicate=predicate))
    B.vstore([T.Ramp(i_0 * 4, 1, 4)], A_load, predicate=predicate)
```
It takes a conservative approach for now, focussing only on expressions produced by the split scheduling primitive, but more complex expressions could be supported in the future.

`TryPredicateBufferAccesses` can be explicitly enabled/disabled with the `tir.enable_buffer_level_predication` pass context option. By default it will be disabled, unless the target supports SVE, in which case it will be enabled by default.

~Note: this commit depends on https://github.com/apache/tvm/pull/16965, so also contains the contents of https://github.com/apache/tvm/pull/16965.~

Co-authored-by: Elen Kalda <elen.kalda@arm.com>
Co-authored-by: Neil Hickey <neil.hickey@arm.com>

### CODE REVIEW

**Comment by lhutton1** [comment]
cc @ekalda @tqchen @Lunderberg @cbalint13 @Anndrey24

File: include/tvm/script/ir_builder/tir/ir.h [hunk 1]
```diff
@@ -411,8 +411,10 @@ Var EnvThread(String thread_tag, DataType dtype = DataType::Int(32));
  * \param buffer The buffer.
  * \param value The value to be stored.
  * \param indices The indices location to be stored.
+ * \param predicate A vector mask of int1 values indicating which lanes of a vector are to be
```
Comment (Lunderberg) [inline include/tvm/script/ir_builder/tir/ir.h:414]: Can we specify that the number of lanes for the `predicate` must be equal to the number of lanes in the `value`?

File: include/tvm/tir/expr.h [hunk 2]
```diff
@@ -675,7 +679,8 @@ class BufferLoadNode : public PrimExprNode {
  */
 class BufferLoad : public PrimExpr {
  public:
-  TVM_DLL explicit BufferLoad(Buffer buffer, Array<PrimExpr> indices, Span span = Span());
+  TVM_DLL explicit BufferLoad(Buffer buffer, Array<PrimExpr> indices,
+                              PrimExpr predicate = PrimExpr(), Span span = Span());
```
Comment (Lunderberg) [inline include/tvm/tir/expr.h:683]: Using `PrimExpr()` as the default would have an undefined object as the predicate.  While this is legal, it is rather unexpected.  While changing `PrimExpr` to be non-nullable would involve more breakage than I'd want to handle at the moment, we should avoid introducing additional cases where `PrimExpr` contains null.

Changing this to `Optional<PrimExpr> predicate = NullOpt` would avoid this issue.

File: include/tvm/tir/stmt.h [hunk 3]
```diff
@@ -261,7 +265,7 @@ class BufferStoreNode : public StmtNode {
 class BufferStore : public Stmt {
  public:
   TVM_DLL explicit BufferStore(Buffer buffer, PrimExpr value, Array<PrimExpr> indices,
-                               Span span = Span());
+                               PrimExpr predicate = PrimExpr(), Span span = Span());
```
Comment (Lunderberg) [inline include/tvm/tir/stmt.h:268]: Same request here, `Optional<PrimExpr> predicate = NullOpt`.

File: python/tvm/tir/buffer.py [hunk 4]
```diff
@@ -141,6 +141,57 @@ def vstore(self, begin, value):
         begin = (begin,) if isinstance(begin, (int, PrimExpr)) else begin
         return _ffi_api.BufferVStore(self, begin, value)  # type: ignore
 
+    def load(self, indices, predicate=None):
```
Comment (Lunderberg) [inline python/tvm/tir/buffer.py:144]: These look very similar to the existing `vload` and `vstore` methods.  Should they be extended instead with a `predicate` argument instead, rather than introducing new methods?

File: python/tvm/tir/expr.py [hunk 5]
```diff
@@ -1093,20 +1093,27 @@ class BufferLoad(PrimExprWithOp):
...
 
     indices : List[PrimExpr]
-        The buffer indices.
+        The buffer indices to load values from.
 
     span : Optional[Span]
         The location of this expression in the source code.
+
+    predicate : Optional[PrimExpr]
+        A vector mask of int1 values indicating which lanes of a vector are to be loaded.
```
Comment (Lunderberg) [inline python/tvm/tir/expr.py:1102]: At first glance, the `int1` reads like a typo.  Can we explicitly call it `boolean values` instead?

File: python/tvm/tir/stmt.py [hunk 6]
```diff
@@ -224,24 +224,29 @@ class BufferStore(Stmt):
     indices : List[PrimExpr]
         The indices location to be stored.
 
+    predicate : Optional[PrimExpr]
+        A vector mask of int1 values indicating which lanes of a vector are to be stored.
```
Comment (Lunderberg) [inline python/tvm/tir/stmt.py:228]: Same nitpick here, using `boolean` instead of `int1`.

File: src/tir/transforms/vectorize_loop.cc [hunk 7]
```diff
@@ -72,6 +72,126 @@ inline PrimExpr BroadcastTo(PrimExpr e, int lanes, bool is_scalable) {
   return Broadcast(e, CreateNewLanes(is_scalable, lanes));
 }
 
+bool EnableBufferLevelPredication() {
```
Comment (Lunderberg) [inline src/tir/transforms/vectorize_loop.cc:75]: I like the fallback, where it checks whether a target supports SVE in order to apply the vectorization if not specified in the pass context.  However, this function should accept a `Target` argument from the calling scope rather than using `Target::Current()`, because the `Target::Current()` may be overridden. 

The override can be either at the function level (`prim_func->GetAttr<Target>(tvm::attr::kTarget)`), or for individual blocks of TIR (an `AttrStmt` with `attr_stmt->attr_key==tvm::attr::kTarget`).

Comment (Lunderberg) [reply src/tir/transforms/vectorize_loop.cc:75]: This will require updating `arith::TargetHasSVE()` as well.  Apologies for not catching it earlier in #16893.

File: tests/python/tir-transform/test_tir_transform_vectorize.py [hunk 8]
```diff
@@ -488,5 +529,170 @@ def main(A: T.Buffer((16,), "float32")):
...
+        for i_0, i_1_s in T.grid(4, 4):
+            if i_0 * 4 + i_1_s < 14:
+                B[i_0 * 4 + i_1_s] = A[i_0 * 4 + i_1_s] + T.float32(1)
+
+    mod = tvm.IRModule.from_expr(before)
+    with tvm.transform.PassContext(config={"tir.enable_buffer_level_predication": False}):
+        with tvm.target.Target("llvm -mtriple=aarch64-linux-gnu -mattr=+sve"):
+            after = tvm.tir.transform.VectorizeLoop()(mod)["main"]
+    tvm.ir.assert_structural_equal(after, expected)
+
```
Comment (Lunderberg) [inline tests/python/tir-transform/test_tir_transform_vectorize.py:695]: Can we also add test cases for the target override both in the function attributes (`T.func_attr({'target': T.target(...)})`) and in an attr stmt (`T.attr(T.target(...), "target", 0)`)?

([Link](https://github.com/apache/tvm/blob/main/tests/python/tir-transform/test_tir_transform_annotate_device_regions.py#L27) to example for the TVMScript syntax.)

File: src/target/llvm/codegen_llvm.cc [hunk 9]
```diff
@@ -1768,11 +1774,17 @@ llvm::Value* CodeGenLLVM::VisitExpr_(const BufferLoadNode* op) {
...
-                                  bool is_volatile) {
+  auto make_load = [this, &loads](TypedPointer buffer_ptr, int /* subelement_i */,
+                                  llvm::Value* predicate, int alignment, bool is_volatile) {
 #if TVM_LLVM_VERSION >= 110
-    auto load = builder_->CreateAlignedLoad(buffer_ptr.type, buffer_ptr.addr,
-                                            llvm::Align(alignment), is_volatile);
+    llvm::Instruction* load = nullptr;
+    if (predicate != NULL) {
+      load = builder_->CreateMaskedLoad(buffer_ptr.type, buffer_ptr.addr, llvm::Align(alignment),
+                                        predicate);
```
Comment (Lunderberg) [inline src/target/llvm/codegen_llvm.cc:1783]: It looks like this ignores the `is_volatile` argument.  Can we provide `is_volatile` to `CreateMaskedLoad`?  If not, we should specify that these two arguments are incompatible and throw an exception if they are both provided.

File: src/target/llvm/codegen_llvm.cc [hunk 10]
```diff
@@ -1768,11 +1774,17 @@ llvm::Value* CodeGenLLVM::VisitExpr_(const BufferLoadNode* op) {
...
+    } else {
+      load = builder_->CreateAlignedLoad(buffer_ptr.type, buffer_ptr.addr, llvm::Align(alignment),
+                                         is_volatile);
+    }
 #elif TVM_LLVM_VERSION >= 80
     auto load =
```
Comment (Lunderberg) [inline src/target/llvm/codegen_llvm.cc:1789]: The PR only adds `CreateMaskedLoad` when `TVM_LLVM_VERSION >= 110`.  If somebody is using an older version of LLVM, it would silently ignore the predicate for the load/store.  We should either support it, or throw an exception.

It looks like `CreateMaskedLoad` has been supported in LLVM since [this commit](https://github.com/llvm/llvm-project/commit/f1de34b84dea91b5060ee0fafbadaad5deaf199c), so I'd lean toward adding it in the other `#elif` branches.

File: tests/python/codegen/test_target_codegen_aarch64.py [hunk 11]
```diff
@@ -700,5 +700,31 @@ def before(a: T.handle):
     assert "get.active.lane.mask" in ll
 
 
+@pytest.mark.skipif(
```
Comment (Lunderberg) [inline tests/python/codegen/test_target_codegen_aarch64.py:703]: This validates that we have the correct output for architectures that support SVE, but it doesn't test the behavior of other targets that do not (yet) support predicated loads/stores.  While the `VectorizeLoop` pass would only insert a predicated load/store for targets that support it, the predicated load/store could still be generated in hand-written kernels, or through other transforms in the future.

Can we add a test, parametrized over each target tested in CI, which attempts to compile a PrimFunc containing predicated loads/stores?  For each target that supports sve, `tvm.build` should compile without error, and for each target that does not, `tvm.build` should raise an exception.

File: src/tir/ir/stmt.cc [hunk 12]
```diff
@@ -458,7 +458,8 @@ TVM_REGISTER_GLOBAL("tir.Evaluate").set_body_typed([](PrimExpr value, Span span)
 TVM_REGISTER_NODE_TYPE(EvaluateNode);
 
 // BufferStore
-BufferStore::BufferStore(Buffer buffer, PrimExpr value, Array<PrimExpr> indices, Span span) {
+BufferStore::BufferStore(Buffer buffer, PrimExpr value, Array<PrimExpr> indices, PrimExpr predicate,
+                         Span span) {
   ICHECK_EQ(buffer->shape.size(), indices.size())
```
Comment (Lunderberg) [inline src/tir/ir/stmt.cc:463]: We should add an assert here, that if `predicate` is defined, the number of lanes in `value` and in `predicate` are identical.

**Review by Lunderberg — CHANGES_REQUESTED** [review]
I like the functionality overall.  I have a couple of requested changes, which mostly fall under the categories below.

* TIR edge cases, when `Target::Current()` may be overridden.
* Using `Optional<PrimExpr>` instead of `PrimExpr`.
* Validating that `!predicate.defined()` in any target that does not support it.

File: python/tvm/tir/buffer.py (hunk 4 above)
Comment (lhutton1) [reply python/tvm/tir/buffer.py:144]: We did look at reusing these but decided against it since the vload/vstore api doesn't let you specify the "indices" argument of the underlying buffer load/store which can limit expressing more complex (e.g. non-consecutive) accesses

File: src/tir/transforms/vectorize_loop.cc (hunk 7 above)
Comment (lhutton1) [reply src/tir/transforms/vectorize_loop.cc:75]: Thanks, that makes sense. Do you know if there is any general infrastructure for keep track of the current target (which takes into account this override functionality) from within a pass? Otherwise I feel we will be duplicating this functionality in multiple places. I was thinking something similar to: `LexicalOnDeviceMixin` (assuming I understood it correctly)

**Review by lhutton1 — COMMENTED** [review]
Thanks for taking the time to review @Lunderberg, I'm working through the comments but wanted to ask a couple of questions / respond to some of the comments before I continue

File: src/tir/transforms/vectorize_loop.cc (hunk 7 above)
Comment (Lunderberg) [reply src/tir/transforms/vectorize_loop.cc:75]: To my knowledge, there isn't a general utility that would provide it, though I definitely agree on its usefulness.

There's currently a couple similar utilities, such as [`IRMutatorWithAnalyzer`](https://github.com/apache/tvm/blob/main/src/arith/ir_mutator_with_analyzer.h#L45) and [`IRVisitorWithAnalyzer`](https://github.com/apache/tvm/blob/main/src/arith/ir_visitor_with_analyzer.h#L35), which allow context-dependent algebraic simplifications. The easiest way to implement such a utility for target-aware simplifications would be to extract out the context-dependent target tracking from an existing visitor/mutator (e.g. [here](https://github.com/apache/tvm/blob/main/src/tir/transforms/lower_device_kernel_launch.cc#L65), used in the `LowerDeviceKernelLaunch` transform) into a `IRVisitorWithTarget` and `IRMutatorWithTarget`, which is then subclassed again by each transform that requires it.

For long-term maintainability, I like your suggestion of mixins much more than the subclasses, because they could be composed together much more easily.  Implemented with subclasses, a hypothetical transform that requires both the context-dependent target and context-dependent algebraic simplifications would have incorrect behavior from multiple inheritance, and would need to re-implement one or the other functionalities.

Unfortunately, I don't think the current `tir::StmtVisitor` API could be used to support mixins, either through multiple inheritance or through composition.  The main issue is that the implementable virtual functions (e.g. `tir::StmtVisitor::VisitStmt_(const tir::ForNode*)`) aren't granular enough.  In order to compose separate transforms that provide context-dependent information, it would require separate virtual functions for pre-visit and post-visit.  I've toyed around with different variations of the TIR stmt/expr visitor/mutator implementations, but haven't had the time to make anything PR-able.

TL;DR: It doesn't currently exist, would be very useful to have, and a subclass would be sufficient for this PR, though long-term maintainability concerns may appear at some point in the future.

File: python/tvm/tir/buffer.py (hunk 4 above)
Comment (Lunderberg) [reply python/tvm/tir/buffer.py:144]: Ah, I see.  I think this is an unfortunate confusion resulting from the name `begin` for the parameter in `vload` and `vstore`.  This parameter would be better named `indices`, because it is used to set the `indices` argument of the buffer load/store.

While there's some additional logic to handle automatic promotion from a scalar index to a vector index when needed ([link](https://github.com/apache/tvm/blob/main/src/tir/ir/buffer.cc#L410)), a vector index can also be provided directly.

File: src/target/llvm/codegen_llvm.cc (hunk 9 above)
Comment (Lunderberg) [reply src/target/llvm/codegen_llvm.cc:1783]: Thank you, and throwing the exception makes sense.  (And we should probably also add a unit test to verify that the exception is thrown.)

**Comment by Lunderberg** [comment]
No problem, and thank you on the revisions!

File: python/tvm/tir/expr.py (hunk 5 above)
Comment (lhutton1) [reply python/tvm/tir/expr.py]: Done. There was a bit of confusion here since LLVM intrinsics expect predicate masks of `int1` values. However, boolean types are already converted to `int1` in the codegen backend. So I have required `uint1`/boolean types in TVM.

File: src/target/llvm/codegen_llvm.cc (hunk 9 above)
Comment (lhutton1) [reply src/target/llvm/codegen_llvm.cc:1785]: Done

File: src/tir/transforms/vectorize_loop.cc (hunk 7 above)
Comment (lhutton1) [reply src/tir/transforms/vectorize_loop.cc]: Thanks for the detailed description. To save this PR exploding further, I've added the minimum changes necessary for the vectorizer pass. Perhaps we can clean this up and fix the target checks in the analyzer in a separate PR?

File: tests/python/tir-transform/test_tir_transform_vectorize.py (hunk 8 above)
Comment (lhutton1) [reply tests/python/tir-transform/test_tir_transform_vectorize.py:696]: Done

File: tests/python/codegen/test_target_codegen_aarch64.py (hunk 11 above)
Comment (lhutton1) [reply tests/python/codegen/test_target_codegen_aarch64.py:703]: Done

File: src/tir/ir/stmt.cc (hunk 12 above)
Comment (lhutton1) [reply src/tir/ir/stmt.cc:463]: Done

**Review by ekalda — COMMENTED** [review]
Thanks @lhutton1 for all the work on this (it's a lot of work) and @Lunderberg for constructive reviews! I've only got some minor nits.

File: src/tir/ir/stmt.cc [hunk 13]
```diff
@@ -476,29 +477,39 @@ BufferStore::BufferStore(Buffer buffer, PrimExpr value, Array<PrimExpr> indices,
...
 
   ICHECK_EQ(index_lanes * buffer_lanes, value_dtype_lanes)
       << "Cannot store value with " << value_dtype_lanes << ", expected value with "
       << index_lanes * buffer_lanes << " (" << index_lanes << " index lanes * " << buffer_lanes
       << " buffer element lanes)";
 
+  if (predicate.defined()) {
+    DataType predicate_dtype = predicate.value().dtype();
+    int predicate_dtype_lanes = predicate_dtype.get_lanes_or_vscale_factor();
+    ICHECK_EQ(value_dtype_lanes, predicate_dtype_lanes)
```
[the remark this answers is not in the document]
Comment (lhutton1) [reply src/tir/ir/stmt.cc:502]: Good catch, yes we should

File: include/tvm/tir/buffer.h [hunk 14]
```diff
@@ -209,14 +209,20 @@ class Buffer : public ObjectRef {
    * \brief Create an Expr that does a vector load at begin index.
    * \param begin The beginning index
    * \param dtype The data type to be loaded.
+   * \param predicate A vector mask of boolean values indicating which lanes of a vector are to be
+   * stored. The number lanes of the mask must be equal to the number of lanes in value.
```
[the remark this answers is not in the document]
Comment (lhutton1) [reply include/tvm/tir/buffer.h]: Oops thanks, result of copy and pasting

**Review by ekalda — APPROVED** [review]
Thanks @lhutton1 LGTM! 🚀

**Comment by lhutton1** [comment]
@tvm-bot rerun

**Comment by lhutton1** [comment]
friendly ping @Lunderberg if you have some free time

**Review by leandron — APPROVED** [review]
This patch is waiting here for more than a week now. I think we should merge it before it starts creating merge-conflicts, so I'm going to merge it now.

@Lunderberg, I'm sure @lhutton1 will be happy to take any other outstanding comment in follow-up patches, please reply when you can. Thanks.

File: tests/python/codegen/test_target_codegen_aarch64.py [hunk 15]
```diff
@@ -780,5 +780,31 @@ def before(a: T.handle):
...
+    def before(a: T.handle, b: T.handle):
+        A = T.match_buffer(a, (16,), "float32")
+        B = T.match_buffer(b, (16,), "float32")
+        T.func_attr({"global_symbol": "main", "tir.noalias": True})
+        for i_0 in T.serial(T.ceildiv(16, 4 * T.vscale())):
+            for i_1 in T.vectorized(4 * T.vscale()):
+                if i_0 * 4 * T.vscale() + i_1 < 14:
+                    B[i_0 * 4 * T.vscale() + i_1] = A[i_0 * 4 * T.vscale() + i_1] + 1.0
+
+    with tvm.target.Target(target):
```
Comment (tqchen) [inline tests/python/codegen/test_target_codegen_aarch64.py:800]: we should avoid this style of code(this was created for legacy usage), instead, attach the target attribute to the PrimFunc itself

**Comment by Lunderberg** [comment]
Apologies for not getting back to the review, and thank you for making the changes!
